### DISCUSSION TITLE
How can I custom the "trailingSlash" for each page in NextJS?

### DISCUSSION THREAD

**Discussion by daoletrinh** [vercel/next.js#44564] [Help]
Hi,
I have a problem with my project. I have some page like:
/about
/payment
...
**trailingSlash: true**
-> I want apply trailingSlash for all pages except /payment
/about -> /about/
...
**/payment -> /payment**
So, which way I can try to custom the trailingSlash?
I am using Next 12.
Many thank and appreciate for your answer!
Regard,

**Comment by ciruz** [comment]
The easiest way is using a ``Middleware`` => https://nextjs.org/docs/advanced-features/middleware#using-middleware 

where you simple append a trailing slash for yourself, whenever you need depending on the url and if there is none existent.

**Comment by icyJoseph** [comment]
This is rather tricky because upon declaring trailing slash true, then Next.js tries to redirect pathnames to their `+/` form, for example `/payment` requests would be, even through middleware, redirected to `/payment/`. This applies the other way around as well.

If you use middleware to redirect, in this situation you'll run into TOO MANY REDIRECTS.

I am not sure there's a way to hack Next.js itself into doing this. You might need to use some kind of additional proxy layer to serve `/payments` from `/payments/`.

May I ask why do you have this requirement?

## Edit

This applies to Next 12

**Reply by ciruz** [reply]
Hey @icyJoseph check this out => https://github.com/vercel/next.js/blob/d89bdb2b0c4024965b6f2263be9d8eef290ed850/docs/advanced-features/middleware.md?plain=1#L254

**Reply by icyJoseph** [reply]
They are using Next 12 though.

**Reply by ciruz** [reply]
It was added to next.js with ``v12.3.4`` https://github.com/vercel/next.js/compare/v12.3.3...v12.3.4 as experimental, so it should work.

**Reply by daoletrinh** [reply]
hi @icyJoseph , thank for your explaination!
"May I ask why do you have this requirement?"
I have to add some '/' at the end of url for SEO goal.
But some url I don't want to add '/' that SEO don't care (Disallow), we don't want to redirect with '/'.

**Comment by ciruz** [comment]
I just found out about ``skipTrailingSlashRedirect`` where you can handle it in the Middleware, like I suggest before:

> In `v13.1` of Next.js two additional flags were introduced for middleware, `skipMiddlewareUrlNormalize` and `skipTrailingSlashRedirect` to handle advanced use cases.
> `skipTrailingSlashRedirect` allows disabling Next.js default redirects for adding or removing trailing slashes allowing custom handling inside middleware which can allow maintaining the trailing slash for some paths but not others allowing easier incremental migrations.

https://github.com/vercel/next.js/blob/d89bdb2b0c4024965b6f2263be9d8eef290ed850/docs/advanced-features/middleware.md?plain=1#L252
